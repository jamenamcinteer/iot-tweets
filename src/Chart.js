import React, { useState, useEffect } from "react"
import OrdinalFrame from "semiotic/lib/OrdinalFrame"
import { scaleSqrt} from "d3-scale"

const Chart = (() => {
  const [twitterData, setTwitterData] = useState([])
  const [topTwentyWords, setTopTwentyWords] = useState([])
  const [frameProps, setFrameProps] = useState({})
  const [hoveredOn, setHoveredOn] = useState()

  useEffect(() => {
    async function fetchData() {
      try {
        const response = await fetch('/tweets');
        const json = await response.json();
        console.log(json)
        setTwitterData(json)
        localStorage.setItem("twitterData", JSON.stringify(json))
      } catch (error) {
        console.log(error)
      }
    }
    if(!localStorage.getItem("twitterData")) fetchData();
    else {
      setTwitterData(JSON.parse(localStorage.getItem("twitterData")))
    }
  }, [])

  const sort = (a, b) => {
    if(a.count < b.count) return 1
    else if(a.count > b.count) return -1
    else return 0
  }

  

  useEffect(() => {
    const getTopTwentyWords = () => {
      let topTwenty = [];
      let allWords = []
      let wordCounts = {}
  
      // TODO: move this to its own file
      const stopwords = ["with", "to", "of", "the", "there", "are", "more", "this", "on", "its", "in", "and", "for", "gt", "will", "or", "an", "be", "how", "is", "from", "https", "rt", "co", "by", "our", "via", "at", "can", "you", "your", "follow", "as", "what"]
  
      twitterData.map(tweet => {
        // remove urls and @'s
        let words = tweet.text.replace(/https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_+.~#?&//=]*)/g, "")
        words = words.replace(/(^|)@\w+/g, "")
  
        // split tweet into words
        words = tweet.text.split(/\b/);
  
        // for each word, sanitize the word and then add it to wordCounts and allWords
        words.map(word => {
          word = word.trim()
          word = word.toLowerCase()
          if(word.length === 1) word = null
          if(word && word.match(/[^a-zA-Z0-9\d\s:]/g)) word = null
          
          if(word && stopwords.indexOf(word.toLowerCase()) < 0) {
            wordCounts["_" + word] = (wordCounts["_" + word] || 0) + 1
            if(allWords.indexOf(word) < 0) allWords.push(word)
          }
          return false;
        })
        return false;
      })
      console.log(allWords)
      console.log(wordCounts)
      for (let word of allWords) {
          topTwenty.push({ word: word, count: wordCounts["_" + word] })
      }
      topTwenty.sort(sort)
      topTwenty = topTwenty.slice(0, 20)
      setTopTwentyWords(topTwenty)
      return topTwenty;
    }
    // getTopTwentyWords()
    if(twitterData.length > 0 && topTwentyWords.length === 0) console.log(getTopTwentyWords())
    if(twitterData.length > 0) {
      let r = -12
      let g = 207
      let b = 255
      // max r is 255; each "step" between 0 and 255 is 12.75 points for 20 steps
      // max g is 255; each "step" between 209 and 255 is 2.3 points for 20 steps
      let data = topTwentyWords.map(word => {
        r += 12
        g += 2
        return {
          word: word.word,
          count: word.count,
          color: `rgb(${r}, ${g}, ${b})`
        }
      })
      // let annotations = topTwentyWords.map(word => {
      //   return {
      //     color: "#FFFFFF",
      //     // type: AnnotationCalloutElbow,
      //     // type: "custom",
      //     word: word.word,
      //     count: word.count,
      //     note: { title: word.word, wrap: 180 }
      //   }
      // })
      let newFrameProps = {
        data: data,
        size: [400, 400],
        margin: 50,
        type: { type: "bar", innerRadius: 200 },
        projection: "radial",
        dynamicColumnWidth: "count",
        oAccessor: "word",
        rAccessor: "count",
        rScaleType: scaleSqrt(),
        // axes: true,
        style: d=> ({ fill: d.color }),
        oPadding: 10,
        pieceHoverAnnotation: true,
        tooltipContent: d => {
          return (
          <div></div>
        )},
        customHoverBehavior: d => {
          setHoveredOn(d)
        }
      }
      setFrameProps(newFrameProps)
    }
  }, [twitterData, topTwentyWords])

  useEffect(() => {
    if(hoveredOn) {
      console.log(hoveredOn)
    }
    
  }, [hoveredOn])
  

  return (
  <div>
    <OrdinalFrame {...frameProps} />
  </div>)
})

export default Chart;