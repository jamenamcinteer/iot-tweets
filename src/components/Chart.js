import React, { useState, useEffect } from "react"
import OrdinalFrame from "semiotic/lib/OrdinalFrame"
import { scaleSqrt} from "d3-scale"

const Chart = (({ twitterData, topTwentyWords }) => {
  const [frameProps, setFrameProps] = useState({})
  const [hoveredOn, setHoveredOn] = useState()

  useEffect(() => {
    if(twitterData.length > 0) {
      let r = -12
      let g = 207
      let b = 255
      // max r is 255; each "step" between 0 and 255 is 12.75 points for 20 steps
      // max g is 255; each "step" between 209 and 255 is 2.3 points for 20 steps
      let data = topTwentyWords.map(word => {
        r += 12
        g += 2
        return {
          word: word.word,
          count: word.count,
          color: `rgb(${r}, ${g}, ${b})`
        }
      })
      let newFrameProps = {
        data: data,
        size: [400, 400],
        margin: 50,
        type: { type: "bar", innerRadius: 200 },
        projection: "radial",
        dynamicColumnWidth: "count",
        oAccessor: "word",
        rAccessor: "count",
        rScaleType: scaleSqrt(),
        // axes: true,
        style: d=> ({ fill: d.color }),
        oPadding: 10,
        pieceHoverAnnotation: true,
        tooltipContent: d => {
          return (
          <div></div>
        )},
        customHoverBehavior: d => {
          setHoveredOn(d)
        }
      }
      setFrameProps(newFrameProps)
    }
  }, [twitterData, topTwentyWords])

  useEffect(() => {
    if(hoveredOn) {
      console.log(hoveredOn)
    }
    
  }, [hoveredOn])


  return (
    <OrdinalFrame {...frameProps} />
  )
})

export default Chart;