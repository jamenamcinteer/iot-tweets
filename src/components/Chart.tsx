import React, { useState, useEffect } from "react"
import OrdinalFrame from "semiotic/lib/OrdinalFrame"
import { scaleSqrt } from "d3-scale"
import { IFrameProps, ITopTwentyWords, IChartData } from "../interfaces/interfaces"

const Chart = (({ twitterData, topTwentyWords, handleHover, handleCycleOn }) => {
  const [frameProps, setFrameProps] = useState<IFrameProps | undefined>(undefined)
  const [hoveredOn, setHoveredOn] = useState()
  const [cycleTweet, setCycleTweet] = useState(0)
  const [cycleOn, setCycleOn] = useState(topTwentyWords[0])

  useEffect(() => {
    if(twitterData.length > 0) {
      let r: number = -12
      let g: number = 207
      let b: number = 255
      // max r is 255; each "step" between 0 and 255 is 12.75 points for 20 steps
      // max g is 255; each "step" between 209 and 255 is 2.3 points for 20 steps
      let data: Array<IChartData> = topTwentyWords.map((word: ITopTwentyWords, index: number) => {
        r += 12
        g += 2
        return {
          word: word.word,
          count: word.count,
          color: `rgb(${r}, ${g}, ${b})`,
          fillOpacity: index === 0 ? 1 : 0.4
        }
      })
      let newFrameProps = {
        data: data,
        size: [400, 400],
        margin: 50,
        type: { type: "bar", innerRadius: 200 },
        projection: "radial",
        dynamicColumnWidth: "count",
        oAccessor: "word",
        rAccessor: "count",
        rScaleType: scaleSqrt(),
        // axes: true,
        style: d=> ({ fill: d.color, fillOpacity: d.fillOpacity }),
        oPadding: 10,
        pieceHoverAnnotation: true,
        tooltipContent: () => {
          return (<div></div>)
        },
        customHoverBehavior: d => {
          setHoveredOn(d)
        }
      }
      setFrameProps(newFrameProps)
    }
  }, [twitterData, topTwentyWords])

  useEffect(() => {
    let index: number | undefined;
    if(frameProps && frameProps.data && hoveredOn) {
      for(let i=0;i<frameProps.data.length;i++) {
        if(frameProps.data[i].word === hoveredOn.word) {
          index = i
        }
      }
    }
    handleHover(
      hoveredOn
      ? {
          word: hoveredOn.word,
          count: hoveredOn.count,
          index: index
        }
      : undefined
    )
  }, [hoveredOn, handleHover, frameProps])

  useEffect(() => {
    handleCycleOn(cycleOn)
  }, [cycleOn, handleCycleOn])

  useEffect(() => {
    setCycleOn({
      ...topTwentyWords[cycleTweet],
      index: cycleTweet
    })
    
  }, [cycleTweet, topTwentyWords])

  useEffect(() => {
    let timerCount = 0;
    const timer = setInterval(() => {
      timerCount++;
      setCycleTweet(timerCount)
      if(timerCount >= 20) clearInterval(timer)
    }, 3000)
    return () => clearInterval(timer)
  }, [])

  return (
    <OrdinalFrame {...frameProps} />
  )
})

export default Chart;