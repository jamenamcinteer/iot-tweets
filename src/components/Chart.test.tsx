import React from 'react';
import { cleanup, render, wait, fireEvent } from "@testing-library/react";
import "jest-styled-components";
import { mockSuccessResponseTwitter } from "../__mocks__/mockTwitterApiResponse"
import Chart from './Chart';

afterEach(cleanup);

const topTwentyWords = [
  { "word": "ai", "count": 50},
  { "word": "startup", "count": 49},
  { "word": "bigdata", "count": 48},
  { "word": "5g", "count": 47},
  { "word": "cybersecurity", "count": 46},
  { "word": "security", "count": 45},
  { "word": "iot", "count": 44},
  { "word": "2019", "count": 43},
  { "word": "py", "count": 42},
  { "word": "datascience", "count": 41},
  { "word": "machinelearning", "count": 40},
  { "word": "vr", "count": 39},
  { "word": "iot", "count": 38},
  { "word": "bigdata", "count": 37},
  { "word": "analytics", "count": 36},
  { "word": "business", "count": 35},
  { "word": "infosec", "count": 34},
  { "word": "ar", "count": 33},
  { "word": "blockchain", "count": 32},
  { "word": "robotics", "count": 32},
  { "word": "data", "count": 31}
]

it('renders as expected', () => {
  const handleHover = jest.fn()
  const handleCycleOn = jest.fn()

  const { asFragment } = render(<Chart twitterData={mockSuccessResponseTwitter} topTwentyWords={topTwentyWords} handleHover={handleHover} handleCycleOn={handleCycleOn} />)
  expect(asFragment()).toMatchSnapshot()
})