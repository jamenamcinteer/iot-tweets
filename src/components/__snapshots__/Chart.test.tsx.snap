// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<DocumentFragment>
  <div
    class=" radial frame ordinalframe"
    style="background: none;"
  >
    <div
      class="frame-elements"
      style="height: 400px; width: 400px;"
    >
      <div
        class="visualization-layer"
        style="position: absolute;"
      >
        <svg
          class="visualization-layer"
          height="400"
          style="position: absolute;"
          width="400"
        >
          <defs>
            <filter
              id="paintyFilterHeavy"
            >
              <fegaussianblur
                color-interpolation-filters="sRGB"
                id="gaussblurrer"
                in="SourceGraphic"
                result="blur"
                stdDeviation="4"
              />
              <fecolormatrix
                in="blur"
                mode="matrix"
                result="gooey"
                values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 34 -7"
              />
            </filter>
            <filter
              id="paintyFilterLight"
            >
              <fegaussianblur
                color-interpolation-filters="sRGB"
                id="gaussblurrer"
                in="SourceGraphic"
                result="blur"
                stdDeviation="2"
              />
              <fecolormatrix
                in="blur"
                mode="matrix"
                result="gooey"
                values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 34 -7"
              />
            </filter>
            <clippath
              id="matte-clip-undefined"
            />
          </defs>
          <g
            aria-label="Visualization with no title. Use arrow keys to navigate elements."
            class="data-visualization"
            role="group"
            transform="translate(50,50)"
          >
            <g
              aria-label="21 bars in a bar chart"
              class="pieces"
              role="group"
              tabindex="0"
            >
              <g
                aria-label="2019 bar value 43"
                class=""
              >
                <path
                  aria-label="2019 bar value 43"
                  class=""
                  d="M4.427219454297229,199.95099331562093A200,200,0,0,1,-38.560531793985646,196.24751052628673L-31.70394799326878,161.35205034968658A164.43729647013686,164.43729647013686,0,0,0,3.639999989723155,164.39700383669557Z"
                  role="img"
                  style="fill: rgb(84, 223, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="ai bar value 50"
                class=""
              >
                <path
                  aria-label="ai bar value 50"
                  class=""
                  d="M20.905692653530693,-198.90437907365467A200,200,0,0,1,72.90278605449895,-186.23958705251673L58.9243149311912,-150.52977635084864A161.6517505576315,161.6517505576315,0,0,0,16.897209070315263,-160.76620535417499Z"
                  role="img"
                  style="fill: rgb(0, 209, 255); fill-opacity: 1;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="startup bar value 49"
                class=""
              >
                <path
                  aria-label="startup bar value 49"
                  class=""
                  d="M91.97075479875,-177.59893091386616A200,200,0,0,1,134.67423758976602,-147.86091346131764L109.11116157513041,-119.79481976700033A162.03716988173517,162.03716988173517,0,0,0,74.5134040973823,-143.88814069652338Z"
                  role="img"
                  style="fill: rgb(12, 211, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="bigdata bar value 48"
                class=""
              >
                <path
                  aria-label="bigdata bar value 48"
                  class=""
                  d="M149.39215208651427,-132.97362480943272A200,200,0,0,1,196.1022713690291,-39.29248228233577L159.26106960855174,-31.910710223666314A162.42654253489104,162.42654253489104,0,0,0,121.32625372629562,-107.99223063063984Z"
                  role="img"
                  style="fill: rgb(24, 213, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="bigdata bar value 37"
                class=""
              >
                <path
                  aria-label="bigdata bar value 37"
                  class=""
                  d="M121.32625372629562,-107.99223063063984A162.42654253489104,162.42654253489104,0,0,1,159.26106960855174,-31.910710223666314L147.0767035267718,-29.469361711751827A150,150,0,0,0,112.04411406488572,-99.73021860707455Z"
                  role="img"
                  style="fill: rgb(156, 235, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="5g bar value 47"
                class=""
              >
                <path
                  aria-label="5g bar value 47"
                  class=""
                  d="M199.13518539889438,-18.578964883651906A200,200,0,0,1,197.66441228141537,30.47589403844689L160.91859084550046,24.810424227713924A162.81999272221063,162.81999272221063,0,0,0,162.11594718692024,-15.125134635712053Z"
                  role="img"
                  style="fill: rgb(36, 215, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="cybersecurity bar value 46"
                class=""
              >
                <path
                  aria-label="cybersecurity bar value 46"
                  class=""
                  d="M193.39598757842234,50.970501160639124A200,200,0,0,1,175.87605410291087,95.2240179429012L143.53038234598472,77.7112027761789A163.21765129208592,163.21765129208592,0,0,0,157.82819430931758,41.59642742310029Z"
                  role="img"
                  style="fill: rgb(48, 217, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="security bar value 45"
                class=""
              >
                <path
                  aria-label="security bar value 45"
                  class=""
                  d="M164.95896641457696,113.08642447011255A200,200,0,0,1,134.67423758976602,147.86091346131764L110.17676229717188,120.96475916272976A163.61965624455004,163.61965624455004,0,0,0,134.9526468960468,92.51580948862544Z"
                  role="img"
                  style="fill: rgb(60, 219, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="iot bar value 44"
                class=""
              >
                <path
                  aria-label="iot bar value 44"
                  class=""
                  d="M118.48080396358836,161.12820700343485A200,200,0,0,1,25.30353674302171,198.39287040691394L20.75220898304817,162.70809686580654A164.02615329077491,164.02615329077491,0,0,0,97.16975256472898,132.14619990706558Z"
                  role="img"
                  style="fill: rgb(72, 221, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="iot bar value 38"
                class=""
              >
                <path
                  aria-label="iot bar value 38"
                  class=""
                  d="M97.16975256472898,132.14619990706558A164.02615329077491,164.02615329077491,0,0,1,20.75220898304817,162.70809686580654L19.0902885823758,149.67777774379172A150.89027890648987,150.89027890648987,0,0,0,89.38800777565498,121.56340047225459Z"
                  role="img"
                  style="fill: rgb(144, 233, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="py bar value 42"
                class=""
              >
                <path
                  aria-label="py bar value 42"
                  class=""
                  d="M-58.86274386157868,191.14177299869903A200,200,0,0,1,-97.18664904585849,174.79918548791105L-80.10767419165198,144.0810681045745A164.85324883225962,164.85324883225962,0,0,0,-48.51857280381196,157.55171133196905Z"
                  role="img"
                  style="fill: rgb(96, 225, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="datascience bar value 41"
                class=""
              >
                <path
                  aria-label="datascience bar value 41"
                  class=""
                  d="M-114.9257406530674,163.6828461853712A200,200,0,0,1,-145.32290958603588,137.4090679302092L-120.09062590501995,113.55085732710835A165.27418319259897,165.27418319259897,0,0,0,-94.9712895712009,135.2627435296352Z"
                  role="img"
                  style="fill: rgb(108, 227, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="machinelearning bar value 40"
                class=""
              >
                <path
                  aria-label="machinelearning bar value 40"
                  class=""
                  d="M-158.88997419172205,121.4659462621269A200,200,0,0,1,-179.30573467611887,88.59714167092106L-148.55505487122642,73.4028572266877A165.70028297149824,165.70028297149824,0,0,0,-131.640568424512,100.63470833517613Z"
                  role="img"
                  style="fill: rgb(120, 229, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="vr bar value 39"
                class=""
              >
                <path
                  aria-label="vr bar value 39"
                  class=""
                  d="M-187.58440216926243,69.36924435800358A200,200,0,0,1,-197.18758045716297,33.42241932370967L-163.79558232297344,27.762623909059492A166.1317431282711,166.1317431282711,0,0,0,-155.81861858027102,57.62216742343061Z"
                  role="img"
                  style="fill: rgb(132, 231, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="analytics bar value 36"
                class=""
              >
                <path
                  aria-label="analytics bar value 36"
                  class=""
                  d="M-199.6009603899361,12.627613052955159A200,200,0,0,1,-198.98600509658428,-20.11391994868551L-166.6114112077537,-16.841428551438398A167.46043132720158,167.46043132720158,0,0,0,-167.12631460111186,10.573127642404359Z"
                  role="img"
                  style="fill: rgb(168, 237, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="business bar value 35"
                class=""
              >
                <path
                  aria-label="business bar value 35"
                  class=""
                  d="M-195.79346179789124,-40.803435115167794A200,200,0,0,1,-187.04392754228124,-70.81362276820639L-157.03792227243522,-59.45354299507731A167.9155526040126,167.9155526040126,0,0,0,-164.38383667022768,-34.25765677752686Z"
                  role="img"
                  style="fill: rgb(180, 239, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="infosec bar value 34"
                class=""
              >
                <path
                  aria-label="infosec bar value 34"
                  class=""
                  d="M-178.6172421700997,-89.9771126428713A200,200,0,0,1,-163.28164088971008,-115.49504642348845L-137.46454662465177,-97.23367616523308A168.3772233983162,168.3772233983162,0,0,0,-150.37537643833014,-75.75048198102101Z"
                  role="img"
                  style="fill: rgb(192, 241, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="ar bar value 33"
                class=""
              >
                <path
                  aria-label="ar bar value 33"
                  class=""
                  d="M-150.31464725880275,-131.92993147675665A200,200,0,0,1,-130.2014231725418,-151.81432542367236L-109.91977488862285,-128.16600670581155A168.84573487796385,168.84573487796385,0,0,0,-126.89993539667232,-111.37903116296195Z"
                  role="img"
                  style="fill: rgb(204, 243, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="blockchain bar value 32"
                class=""
              >
                <path
                  aria-label="blockchain bar value 32"
                  class=""
                  d="M-113.61924801465102,-164.59242534389372A200,200,0,0,1,-90.59857663258377,-178.30282642782151L-76.70138936926432,-150.95242137128776A169.32140044610517,169.32140044610517,0,0,0,-96.19085095737033,-139.34509981024547Z"
                  role="img"
                  style="fill: rgb(216, 245, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="robotics bar value 32"
                class=""
              >
                <path
                  aria-label="robotics bar value 32"
                  class=""
                  d="M-71.4645477075257,-186.79619487815802A200,200,0,0,1,-45.85420166365676,-194.6725255134599L-38.82048821014244,-164.81112324159585A169.32140044610517,169.32140044610517,0,0,0,-60.50238650042873,-158.14296657386646Z"
                  role="img"
                  style="fill: rgb(228, 247, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
              <g
                aria-label="data bar value 31"
                class=""
              >
                <path
                  aria-label="data bar value 31"
                  class=""
                  d="M-25.254187616783813,-198.39915828404182A200,200,0,0,1,-3.6739403974420595e-14,-200L-3.1192591300867464e-14,-169.8045581936221A169.8045581936221,169.8045581936221,0,0,0,-21.44138085403409,-168.4454070920411Z"
                  role="img"
                  style="fill: rgb(240, 249, 255); fill-opacity: 0.4;"
                  tabindex="-1"
                  transform="translate(150,150)"
                />
              </g>
            </g>
          </g>
          <g
            aria-hidden="true"
            class="foreground-graphics"
          />
        </svg>
      </div>
      <div
        class="interaction-layer"
        style="position: absolute; background: none; pointer-events: none;"
      >
        <svg
          height="400"
          style="background: none; pointer-events: none;"
          width="400"
        >
          <g
            class="interaction-overlay"
            style="pointer-events: all;"
            transform="translate(50,50)"
          >
            <g
              class="interaction-regions"
            >
              <g
                class=""
              >
                <path
                  class=""
                  d="M4.427219454297229,199.95099331562093A200,200,0,0,1,-38.560531793985646,196.24751052628673L-31.70394799326878,161.35205034968658A164.43729647013686,164.43729647013686,0,0,0,3.639999989723155,164.39700383669557Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M20.905692653530693,-198.90437907365467A200,200,0,0,1,72.90278605449895,-186.23958705251673L58.9243149311912,-150.52977635084864A161.6517505576315,161.6517505576315,0,0,0,16.897209070315263,-160.76620535417499Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M91.97075479875,-177.59893091386616A200,200,0,0,1,134.67423758976602,-147.86091346131764L109.11116157513041,-119.79481976700033A162.03716988173517,162.03716988173517,0,0,0,74.5134040973823,-143.88814069652338Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M149.39215208651427,-132.97362480943272A200,200,0,0,1,196.1022713690291,-39.29248228233577L159.26106960855174,-31.910710223666314A162.42654253489104,162.42654253489104,0,0,0,121.32625372629562,-107.99223063063984Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M121.32625372629562,-107.99223063063984A162.42654253489104,162.42654253489104,0,0,1,159.26106960855174,-31.910710223666314L147.0767035267718,-29.469361711751827A150,150,0,0,0,112.04411406488572,-99.73021860707455Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M199.13518539889438,-18.578964883651906A200,200,0,0,1,197.66441228141537,30.47589403844689L160.91859084550046,24.810424227713924A162.81999272221063,162.81999272221063,0,0,0,162.11594718692024,-15.125134635712053Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M193.39598757842234,50.970501160639124A200,200,0,0,1,175.87605410291087,95.2240179429012L143.53038234598472,77.7112027761789A163.21765129208592,163.21765129208592,0,0,0,157.82819430931758,41.59642742310029Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M164.95896641457696,113.08642447011255A200,200,0,0,1,134.67423758976602,147.86091346131764L110.17676229717188,120.96475916272976A163.61965624455004,163.61965624455004,0,0,0,134.9526468960468,92.51580948862544Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M118.48080396358836,161.12820700343485A200,200,0,0,1,25.30353674302171,198.39287040691394L20.75220898304817,162.70809686580654A164.02615329077491,164.02615329077491,0,0,0,97.16975256472898,132.14619990706558Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M97.16975256472898,132.14619990706558A164.02615329077491,164.02615329077491,0,0,1,20.75220898304817,162.70809686580654L19.0902885823758,149.67777774379172A150.89027890648987,150.89027890648987,0,0,0,89.38800777565498,121.56340047225459Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-58.86274386157868,191.14177299869903A200,200,0,0,1,-97.18664904585849,174.79918548791105L-80.10767419165198,144.0810681045745A164.85324883225962,164.85324883225962,0,0,0,-48.51857280381196,157.55171133196905Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-114.9257406530674,163.6828461853712A200,200,0,0,1,-145.32290958603588,137.4090679302092L-120.09062590501995,113.55085732710835A165.27418319259897,165.27418319259897,0,0,0,-94.9712895712009,135.2627435296352Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-158.88997419172205,121.4659462621269A200,200,0,0,1,-179.30573467611887,88.59714167092106L-148.55505487122642,73.4028572266877A165.70028297149824,165.70028297149824,0,0,0,-131.640568424512,100.63470833517613Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-187.58440216926243,69.36924435800358A200,200,0,0,1,-197.18758045716297,33.42241932370967L-163.79558232297344,27.762623909059492A166.1317431282711,166.1317431282711,0,0,0,-155.81861858027102,57.62216742343061Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-199.6009603899361,12.627613052955159A200,200,0,0,1,-198.98600509658428,-20.11391994868551L-166.6114112077537,-16.841428551438398A167.46043132720158,167.46043132720158,0,0,0,-167.12631460111186,10.573127642404359Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-195.79346179789124,-40.803435115167794A200,200,0,0,1,-187.04392754228124,-70.81362276820639L-157.03792227243522,-59.45354299507731A167.9155526040126,167.9155526040126,0,0,0,-164.38383667022768,-34.25765677752686Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-178.6172421700997,-89.9771126428713A200,200,0,0,1,-163.28164088971008,-115.49504642348845L-137.46454662465177,-97.23367616523308A168.3772233983162,168.3772233983162,0,0,0,-150.37537643833014,-75.75048198102101Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-150.31464725880275,-131.92993147675665A200,200,0,0,1,-130.2014231725418,-151.81432542367236L-109.91977488862285,-128.16600670581155A168.84573487796385,168.84573487796385,0,0,0,-126.89993539667232,-111.37903116296195Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-113.61924801465102,-164.59242534389372A200,200,0,0,1,-90.59857663258377,-178.30282642782151L-76.70138936926432,-150.95242137128776A169.32140044610517,169.32140044610517,0,0,0,-96.19085095737033,-139.34509981024547Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-71.4645477075257,-186.79619487815802A200,200,0,0,1,-45.85420166365676,-194.6725255134599L-38.82048821014244,-164.81112324159585A169.32140044610517,169.32140044610517,0,0,0,-60.50238650042873,-158.14296657386646Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
              <g
                class=""
              >
                <path
                  class=""
                  d="M-25.254187616783813,-198.39915828404182A200,200,0,0,1,-3.6739403974420595e-14,-200L-3.1192591300867464e-14,-169.8045581936221A169.8045581936221,169.8045581936221,0,0,0,-21.44138085403409,-168.4454070920411Z"
                  style="opacity: 0;"
                  transform="translate(150,150)"
                  type="frame-hover"
                />
              </g>
            </g>
          </g>
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
